test/e2e/storage/volume_provisioning.go

```diff
@@ -135,13 +135,19 @@ func testDynamicProvisioning(t storageClassTest, client clientset.Interface, cla
 	By(fmt.Sprintf("deleting claim %q/%q", claim.Namespace, claim.Name))
 	framework.ExpectNoError(client.CoreV1().PersistentVolumeClaims(claim.Namespace).Delete(claim.Name, nil))
 
-	// Wait for the PV to get deleted. Technically, the first few delete
+	// Wait for the PV to get deleted if reclaim policy is Delete. (If it's
+	// Retain, there's no use waiting because the PV won't be auto-deleted and
+	// it's expected for the caller to do it.) Technically, the first few delete
 	// attempts may fail, as the volume is still attached to a node because
 	// kubelet is slowly cleaning up the previous pod, however it should succeed
 	// in a couple of minutes. Wait 20 minutes to recover from random cloud
 	// hiccups.
-	By(fmt.Sprintf("deleting the claim's PV %q", pv.Name))
-	framework.ExpectNoError(framework.WaitForPersistentVolumeDeleted(client, pv.Name, 5*time.Second, 20*time.Minute))
+	if pv.Spec.PersistentVolumeReclaimPolicy == v1.PersistentVolumeReclaimDelete {
```

Comment:
Sorry for lack of knowledge here, but  isn't the default reclaim policy "Delete" and  the pv automatically deleted as part of dynamic provisioning when the pvc is deleted. If this is true then the e2e test should not have to explicitly delete the pv, but just wait some and check that the pv was deleted (for reclaimPolicy == "Delete"). And, the `if` here should test for reclaimPolicy == "Retain" ?

Reply:
the if statement is correct but confusingly inverted because it's saying: if reclaim policy is delete, we expect the pv to be deleted by kube and so we wait for it; else if reclaim policy is retain, there's nothing to do or check for because the job of the kube dynamic provisioner is done.

Reply:
ack. makes sense and I read the code below too quickly...